File: src/main/java/leet_code/LeafSimilarTrees.java
package leet_code;

import java.util.ArrayList;

public class LeafSimilarTrees {

    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode() {
        }

        TreeNode(int val) {
            this.val = val;
        }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    public static void main(String[] args) {

    }

    public boolean leafSimilar(TreeNode root1, TreeNode root2) {
        ArrayList<Integer> leaves1 = new ArrayList<>();
        ArrayList<Integer> leaves2 = new ArrayList<>();
        dfs(root1, leaves1);
        dfs(root2, leaves2);
        return leaves1.equals(leaves2);
    }

    private void dfs(TreeNode root, ArrayList<Integer> leaves) {
        if (root != null) {
            if (root.left == null && root.right == null) {
                leaves.add(root.val);
            }

            // Depth first search
            dfs(root.left, leaves);
            dfs(root.right, leaves);
        }
    }
}

File: src/main/java/leet_code/LeafSimilarTrees.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func leafSimilar(root1 *TreeNode, root2 *TreeNode) bool {

	var l1 []int

	var l2 []int

	dfs(&l1, root1)
	dfs(&l2, root2)
	return testEq(l1, l2)
}

func dfs(l *[]int, root *TreeNode) {

	if root != nil {
		if root.Left == nil && root.Right == nil {
			*l = append(*l, root.Val)
		}
	}

	dfs(l, root.Left)
	dfs(l, root.Right)

}

func testEq(a, b []int) bool {
	if len(a) != len(b) {
		return false
	}
	for i := range a {
		if a[i] != b[i] {
			return false
		}
	}
	return true
}
